### ISSUE TITLE
Improve `mock_reduce` on test for `MapReduce`

### ISSUE THREAD

**Issue by qh681248** [gchq/coreax#799]
### What's the problem?

At the moment `mock_reduce` choses first n elements, this can be a problem when `_jit_tree` adds `0` valued `0` weighted padding, 

The data `[10,15,20,30,40,50,60,70,80,90,100,110,120,130,140,150,160,170,180,190,200]` when reduced using `mock_reduce` for instance (with coreset size of `3` and leaf size of `6) returns `[0,0,0]`, which means it is not a padding invariant solver and should not be used on `MapReduce` tests. Furthermore, it returns a `Coreset` object and not a `Coresubset` object. 

This solver cannot be used to test diverse selection (the indexing bug that was present before) 

### How can we reproduce the issue?

`
import sys
from unittest.mock import MagicMock

class _ExplicitPaddingInvariantSolver(ExplicitSizeSolver, PaddingInvariantSolver):
    pass
base_solver = MagicMock(_ExplicitPaddingInvariantSolver)
base_solver.coreset_size = 3

def mock_reduce(
    dataset: Data, solver_state: None = None
) -> tuple[Coreset[Data], None]:
    indices = jnp.arange(base_solver.coreset_size)
    return Coreset(dataset[indices], dataset), solver_state

base_solver.reduce = mock_reduce

dataset = Data([10,15,20,30,40,50,60,70,80,90,100,110,120,130,140,150,160,170,180,190,200])
coresubset, _ = MapReduce(base_solver=base_solver,
            leaf_size=6).reduce(dataset)
`

### Python version

3.11

### Package version

0.2.1

### Operating system

Windows
